cupy/cuda/cub.pyx

```diff
@@ -60,13 +65,21 @@ def reduce_sum(core.ndarray x, out=None):
     ws = core.ndarray(ws_size, numpy.int8)
     ws_ptr = <void *>ws.data.ptr
     cub_reduce_sum(x_ptr, y_ptr, x.size, ws_ptr, ws_size, dtype_id)
+    if keepdims:
+        y = y.reshape((1,))
     if out is not None:
         out[...] = y
         y = out
     return y
 
 
-def can_use_reduce_sum(x_dtype, dtype=None):
+cpdef bint _cub_axis_compatible(axis, Py_ssize_t ndim):
+    if ((axis is None) or ndim == 1 or axis == tuple(range(ndim))):
+        return True
+    return False
+
+
+def can_use_reduce_sum(x_dtype, Py_ssize_t ndim, dtype=None, axis=None):
```

Comment:
Just curious, can `ndim` be signed? I thought `size_t` is sufficient.

Reply:
I wasn't sure what CuPy preferred for integers and think I saw ndim with type `Py_ssize_t` in one of the other Cython files, so I used that. I am pretty sure ndim is only ever in the range [0, 32] where I think the upper limit is a hardcoded constant in NumPy.

Reply:
[Here is the definition with a max of 32](https://github.com/numpy/numpy/blob/2d9601b580adc6b5e5564f512c7192563b0dd485/numpy/core/include/numpy/ndarraytypes.h#L31-L40)

In numpy/core/src/*.c, it looks like `int ndim` is most commonly used, although I did see `npy_intp` instead in a few places.

Reply:
It's funny that there's a hardcoded 32...